### PR TITLE
bpfd: Unify the "run as root" and "run as bpfd user" codepaths

### PR DESCRIPTION
**Pull request by Billy99**
* Remove the bpfd user and user group
  * Remove the User/Group from bpfd.service
  * Remove the directory management from bpfd.service
  * Add directory mode setting to bpfd main.rs with directory creation.
  * Update docs now that bpfctl requires sudo for systemd deployment.
  * Simplify install scripts since user is no longer created.
* Document the current Linux capabilities and why they are needed.
* Add helptext to "cargo xtask".
* Add "cargo xtask copy" to copy binaries so `sudo bpfctl` can be used instead of `sudo ./target/debug/bpfctl`.
* Removed CAP_DAC_OVERRIDE and CAP_PERFMON from list of capabilities required by bpfd. 

Resolves: #742

### CODE REVIEW

**Comment by Billy99** [comment]
/hold

Seeing an issue with CSI deployment, need to determine if it's this PR or not.

**Comment by Billy99** [comment]
cancel /hold

Issue was CSI needs a newer version of K8s and I had an older version of KIND installed.

File: docs/developer-guide/testing.md [hunk 1]
```diff
@@ -70,9 +70,8 @@ directory.  These programs are compiled by executing `cargo xtask build-ebpf
 We also load some tests from local files to test the `load-from-file` option.
 
 The `bpf` directory also contains a script called `build_push_images.sh` that
-can be used to build and push new images to quay if the code is changed.  We
-plan to push the images automatically when code gets merged ([issue
-\#533](<https://github.com/bpfd-dev/bpfd/issues/533>)). However, it's still
+can be used to build and push new images to quay if the code is changed.
+Images automatically when code gets merged, however, it's still
```
Comment (dave-tucker) [inline docs/developer-guide/testing.md:74]:
```suggestion
Images get pushed automatically when code gets merged, however, it's still
```

**Review by dave-tucker — DISMISSED** [review]
Small nit but otherwise LGTM

File: bpfd/src/utils.rs [hunk 2]
```diff
@@ -63,25 +61,18 @@ pub(crate) fn get_ifindex(iface: &str) -> Result<u32, BpfdError> {
 }
 
 pub(crate) async fn set_file_permissions(path: &str, mode: u32) {
```
Comment (astoycos) [inline bpfd/src/utils.rs:63]: You can now use this util here as well -> https://github.com/bpfd-dev/bpfd/blob/main/bpfd/src/serve.rs#L152

File: docs/developer-guide/linux-capabilities.md [hunk 3]
```diff
@@ -11,6 +11,36 @@ required set of capabilities via the `bpfd.service` file using the `AmbientCapab
...
+
+* **CAP_BPF:**
+    * Required to load BPF programs and create BPF maps.
+* **CAP_DAC_READ_SEARCH:**
+    * Required by Tracepoint programs, needed by aya to check the tracefs mount point.
+      For example, trying to read "/sys/kernel/tracing" and "/sys/kernel/debug/tracing".
+* **CAP_NET_ADMIN:**
+    * Required for TC programs to attach/detach to/from a qdisc.
+* **CAP_SETPCAP:**
+    * CAP_SETPCAP is required to allow bpfd to drop Linux Capabilities on spawned threads.
```
Comment (astoycos) [inline docs/developer-guide/linux-capabilities.md:26]:
```suggestion
    * Required to allow bpfd to drop Linux Capabilities on spawned threads.
```

File: docs/developer-guide/linux-capabilities.md [hunk 4]
```diff
@@ -11,6 +11,36 @@ required set of capabilities via the `bpfd.service` file using the `AmbientCapab
...
+* **CAP_SYS_ADMIN:** 
+    * Kprobe (Kprobe and Uprobe) and Tracepoint programs are considered perfmon programs and require CAP_PERFMON and CAP_SYS_ADMIN to load.
+    * TC and XDP programs are considered admin programs and require CAP_NET_ADMIN and CAP_SYS_ADMIN to load.
+* **CAP_SYS_RESOURCE:**
+    * Required by bpfd to call `setrlimit()` on `RLIMIT_MEMLOCK`.
+
+Consider adding:
+* **CAP_DAC_OVERRIDE:**
+    * Found to be needed by MantisNet:
```
Comment (astoycos) [inline docs/developer-guide/linux-capabilities.md:35]: Wait is this a mantisnet specific thing?

File: docs/getting-started/bpfctl-guide.md [hunk 5]
```diff
@@ -16,48 +16,28 @@ sudo ./target/debug/bpfctl list
...
-```console
-sudo cp /$HOME/src/xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o /run/bpfd/examples/.
-sudo cp /$HOME/src/net-ebpf-playground/.output/filter.bpf.o /run/bpfd/examples/.
-sudo chown bpfd:bpfd -R /run/bpfd/examples/
-```
-
-This is only needed if `bpfd` is run as a systemd service.
-When `sudo ./scripts/setup.sh install` is run to launch bpfd as a systemd service, all the
-eBPF object files from the `examples/` directory are copied to `/run/bpfd/examples/`.
+eBPF object files used in the examples are taken from the
```
Comment (astoycos) [inline docs/getting-started/bpfctl-guide.md:29]: Aren't they ALL now defined in `/bpfd/examples/` and `bpfd/integration-test/bpf` now?

File: docs/getting-started/bpfctl-guide.md [hunk 6]
```diff
@@ -310,45 +290,45 @@ Below are some additional examples of `bpfctl load` commands:
 XDP
 
 ```console
-bpfctl load-from-file --path /run/bpfd/examples/xdp_pass_kern.o --name "pass" xdp --iface vethb2795c7 --priority 35
+sudo bpfctl load-from-file --path $HOME/src/xdp-tutorial/basic01-xdp-pass/xdp_pass_kern.o --name "pass" xdp --iface vethb2795c7 --priority 35
```
Comment (astoycos) [inline docs/getting-started/bpfctl-guide.md:293]: you can just do `/bpfd/tests/integration-test/bpf/.output/xdp_pass.bpf.o` 👍

File: docs/getting-started/bpfctl-guide.md [hunk 7]
```diff
@@ -310,45 +290,45 @@ Below are some additional examples of `bpfctl load` commands:
...
 ```
 
 TC
 
 ```console
-bpfctl load-from-file --path /run/bpfd/examples/filter.bpf.o --name classifier tc --direction ingress --iface vethb2795c7 --priority 110
+sudo bpfctl load-from-file --path $HOME/src/net-ebpf-playground/.output/filter.bpf.o --name classifier tc --direction ingress --iface vethb2795c7 --priority 110
```
Comment (astoycos) [inline docs/getting-started/bpfctl-guide.md:299]: The integration test `tc_pass.o` should work for this

File: xtask/src/copy.rs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        panic!("unable to run cargo metadata")
+    }
+    let stdout = String::from_utf8(output.stdout).unwrap();
+    let v: Value = serde_json::from_str(&stdout).unwrap();
+    v["workspace_root"].as_str().unwrap().to_string()
+}
+
+/// Copy the binaries
+pub fn copy(opts: Options) -> Result<(), anyhow::Error> {
+    let root = PathBuf::from(WORKSPACE_ROOT.to_string());
```
Comment (astoycos) [inline xtask/src/copy.rs:30]: Is WORKSPACE_ROOT only used here? If so why make it a pub static ref, and instead just call the `workspace_root` function here?

**Review by astoycos — COMMENTED** [review]
Almost there just some questions, thanks!!

**Review by astoycos — APPROVED** [review]
LGTM

**Pull request merged**